m2m-oauth-server/service/grpc/token.go

```diff
@@ -0,0 +1,285 @@
+)
+
+func setKeyError(key string, err error) error {
+	return fmt.Errorf("failed to set %v: %w", key, err)
+}
+
+func setKeyErrorExt(key, info interface{}, err error) error {
+	return fmt.Errorf("failed to set %v('%v'): %w", key, info, err)
+}
+
+func makeAccessToken(clientCfg *oauthsigner.Client, tokenReq tokenRequest) (jwt.Token, error) {
+	token := jwt.New()
+
+	claims := map[string]interface{}{
+		jwt.JwtIDKey:    tokenReq.id,
+		jwt.SubjectKey:  tokenReq.subject,
+		jwt.AudienceKey: tokenReq.host,
+		jwt.IssuedAtKey: tokenReq.issuedAt,
+		uri.ScopeKey:    tokenReq.scopes,
+		uri.ClientIDKey: clientCfg.ID,
+		jwt.IssuerKey:   tokenReq.host,
+	}
+	for key, val := range claims {
+		if err := token.Set(key, val); err != nil {
+			return nil, setKeyError(key, err)
+		}
+	}
+	if !tokenReq.expiration.IsZero() {
+		if err := token.Set(jwt.ExpirationKey, tokenReq.expiration); err != nil {
+			return nil, setKeyError(jwt.ExpirationKey, err)
+		}
+	}
+	if err := setDeviceIDClaim(token, tokenReq); err != nil {
+		return nil, err
+	}
+	if err := setName(token, tokenReq); err != nil {
+		return nil, err
+	}
+	if err := setOwnerClaim(token, tokenReq); err != nil {
+		return nil, err
+	}
+	if err := setOriginTokenClaims(token, tokenReq); err != nil {
+		return nil, err
+	}
+
+	for k, v := range clientCfg.InsertTokenClaims {
+		if _, ok := token.Get(k); ok {
+			continue
+		}
+		if err := token.Set(k, v); err != nil {
+			return nil, setKeyErrorExt(k, v, err)
+		}
+	}
+
+	return token, nil
+}
+
+func getSubject(clientCfg *oauthsigner.Client, tokenReq tokenRequest) string {
+	if tokenReq.subject != "" {
+		return tokenReq.subject
+	}
+	if tokenReq.owner != "" {
+		return tokenReq.owner
+	}
+	return clientCfg.ID
+}
+
+func setDeviceIDClaim(token jwt.Token, tokenReq tokenRequest) error {
+	if tokenReq.deviceID != "" && tokenReq.deviceIDClaim != "" {
+		return token.Set(tokenReq.deviceIDClaim, tokenReq.deviceID)
+	}
+	return nil
+}
+
+func setOwnerClaim(token jwt.Token, tokenReq tokenRequest) error {
+	if tokenReq.owner != "" && tokenReq.ownerClaim != "" {
+		return token.Set(tokenReq.ownerClaim, tokenReq.owner)
+	}
+	return nil
+}
+
+func setName(token jwt.Token, tokenReq tokenRequest) error {
+	if tokenReq.CreateTokenRequest.GetTokenName() != "" && tokenReq.ownerClaim != "name" {
+		return token.Set("name", tokenReq.CreateTokenRequest.GetTokenName())
+	}
+	return nil
+}
+
+func setOriginTokenClaims(token jwt.Token, tokenReq tokenRequest) error {
+	if len(tokenReq.originalTokenClaims) > 0 {
+		return token.Set(uri.OriginalTokenClaims, tokenReq.originalTokenClaims)
+	}
+	return nil
+}
+
+func getExpirationTime(clientCfg *oauthsigner.Client, tokenReq tokenRequest) time.Time {
+	if clientCfg.AccessTokenLifetime == 0 {
+		if tokenReq.CreateTokenRequest.GetExpiration() > 0 {
+			return time.Unix(tokenReq.CreateTokenRequest.GetExpiration(), 0)
+		}
+		return time.Time{}
+	}
+	if tokenReq.CreateTokenRequest.GetExpiration() == 0 {
```

Comment:
Can we add a test case that goes through lines 117-125? Otherwise we're fine, even if we don't hit 80%, the rest of the code paths are error paths

Reply:
added